cuqi/mcmc/_sampler.py

```diff
@@ -0,0 +1,324 @@
+        """ Validate the target is compatible with the sampler. Called when the target is set. Should raise an error if the target is not compatible. """
+        pass
+
+    @abstractmethod
+    def get_state(self):
+        """ Return the state of the sampler. """
+        pass
+
+    @abstractmethod
+    def set_state(self, state):
+        """ Set the state of the sampler. """
+        pass
+
+
+    # ------------ Public attributes ------------
+
+    @property
+    def initial_point(self):
+        """ Return the initial point of the sampler. This is always the first sample. """
+        return self._samples[0]
+    
+    @property
+    def dim(self):
+        """ Dimension of the target density. """
+        return self.target.dim
+
+    @property
+    def geometry(self):
+        """ Geometry of the target density. """
+        return self.target.geometry
+    
+    @property
+    def target(self) -> cuqi.density.Density:
+        """ Return the target density. """
+        return self._target
+    
+    @target.setter
+    def target(self, value):
+        """ Set the target density. Runs validation of the target. """
+        self._target = value
+        self.validate_target()
+
+
+    # ------------ Public methods ------------
+
+    def get_samples(self) -> Samples:
+        """ Return the samples. The internal data-structure for the samples is a dynamic list so this creates a copy. """
+        return Samples(np.array(self._samples).T, self.target.geometry)
+    
+    def reset(self): # TODO. Issue here. Current point is not reset, and initial point is lost with this reset.
+        self._samples.clear()
+        self._acc.clear()
+    
+    def save_checkpoint(self, path):
+        """ Save the state of the sampler to a file. """
+
+        state = self.get_state()
+
+        with open(path, 'wb') as handle:
+            pkl.dump(state, handle, protocol=pkl.HIGHEST_PROTOCOL)
+
+    def load_checkpoint(self, path):
+        """ Load the state of the sampler from a file. """
+
+        with open(path, 'rb') as handle:
+            state = pkl.load(handle)
+
+        self.set_state(state)
+
+    def sample(self, Ns, batch_size=0, sample_path='./CUQI_samples/') -> 'SamplerNew':
+        """ Sample Ns samples from the target density.
+
+        Parameters
+        ----------
+        Ns : int
+            The number of samples to draw.
+
+        batch_size : int, optional
+            The batch size for saving samples to disk. If 0, no batching is used. If positive, samples are saved to disk in batches of the specified size.
+
+        sample_path : str, optional
+            The path to save the samples. If not specified, the samples are saved to the current working directory under a folder called 'CUQI_samples'.
+
+        """
+
+        # Initialize batch handler
+        if batch_size > 0:
+            batch_handler = _BatchHandler(batch_size, sample_path)
+
+        # Draw samples
+        for _ in progressbar( range(Ns) ):
```

Comment:
Do we want to display some information at sampling? For example, `acc` might be useful. `acc` might be particularly useful for samplers that don't have an auto `tune` yet

Reply:
#378